File: rats/crawler.py
#!/usr/bin/python2

import sys, base64, random, logging, sqlite3
import bs4, feedparser
import requests, requests.exceptions
from PIL import Image

class Storage:
    def __init__(self, name):
        self.conn = sqlite3.connect(name)
        cursor = self.conn.cursor()
        cursor.execute('''create table if not exists images(
            id integer primary key,
            feed_title text,
            feed_url text,
            story_when text,
            story_url text,
            story_title text,
            story_descr text,
            image_type text,
            image_data blob)''')
        self.conn.commit()
        cursor.close()

    def add_story(self, story):
        cursor = None
        try:
            cursor = self.conn.cursor()
            cursor.execute('insert into images values(?, ?, ?, ?, ?, ?, ?, ?, ?)', story)
            self.conn.commit()
        except sqlite3.Error as e:
            logging.error('failed to write story to database: %s', e)
        finally:
            if cursor:
                cursor.close()


class Crawler:
    def __init__(self, storage, max_entries, size_thres):
        self.url_cache = set()
        self.storage = storage
        self.max_entries = max_entries
        self.size_thres = size_thres


    def fetch_images(self, rss_url):
        # fetch the feed
        rss_xml = None
        try:
            resp = requests.get(rss_url)
            if resp.status_code != requests.codes.ok:
                logging.error('non ok %d response from %s', resp.status_code, rss_url)
            return
            rss_xml = resp.body
            logging.info('got %s', rss_url)
        except requests.exceptions.RequestException as e:
            logging.error('failed to retrieve rss from %s: reason:  %s', rss_url, e)
            return

        # parse feed
        feed = feedparser.parse(rss_xml)
        feed_title = feed.feed.title
        feed_url = rss_url
        logging.info('got feed with title %s and %d entries', feed_title, len(feed.entries))

        # fetch images from the entries
        for entry in feed.entries[0: self.max_entries]:
            try:
                entry_html = None
                resp = requests.get(entry.link)
                if resp.status_code != requests.codes.ok:
                    logging.error('non ok %d response from %s', resp.status_code, entry.link)
                    continue
                else:
                    entry_html = resp.body
            except requests.exceptions.RequestException as e:
                logging.error('failed to retrieve entry html from %s: reason:  %s', entry.link, e)
                continue

            soup = bs4.BeautifulSoup(entry_html)
            images = soup.find_all('img')
            for img in images:
                img_url = None
                if "src" in img.attrs:
                    img_url = img["src"]
                if img_url and img_url not in self.url_cache and img_url.startswith('http://'):
                    image_bytes = None
                    try:
                        resp = requests.get(img_url)
                        if resp.status_code != requests.codes.ok:
                            logging.error('non ok %d response from %s', resp.status_code, entry.link)
                            continue
                        else:
                            image_bytes = resp.content
                    except requests.exceptions.RequestException as e:
                        logging.error('failed to retrieve image from %s: reason:  %s', img_url, e)
                        continue

                    image = Image.open(image_bytes)
                    self.url_cache.add(img_url)
                    if image.size < self.size_thres:
                        logging.debug("rejected image %s because of small size %s'", img_url, image.size)
                    else:
                        image_type = unicode(resp.headers['Content-Type'])
                        image_data = u'data:' + image_type + u';base64,' + base64.b64encode(resp.body)
                        pt = entry.published_parsed
                        story = (None, feed_title, feed_url, sqlite3.Timestamp(pt[0], pt[1], pt[2], pt[3], pt[4], pt[5]),
                                 entry.link, entry.title, entry.description, image_type, image_data)
                        self.storage.add_story(story)


def main():
    storage = Storage(sys.argv[1])
    crawler = Crawler(storage, 30, (256, 256))

    with open(sys.argv[2]) as rss_file:
        for rss in rss_file:
            try:
                crawler.fetch_images(rss.strip())
            except Exception as e:
                logging.error('general error %s', e)




if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    main()

File: rats/crawler.rb

require 'base64'
require 'logger'
require 'open-uri'
require 'feedzirra'
require 'sqlite3'
require 'dimensions'

MAX_ENTRIES = 2

$logger = Logger.new(STDOUT)
$logger.level = Logger::DEBUG
$logger.progname = 'rats'



class Storage
  def initialize(name)
    @db = SQLite3::Database.new name
    @db.execute %q{create table if not exists images(
            id integer primary key,
            feed_title text,
            feed_url text,
            story_when text,
            story_url text,
            story_title text,
            story_descr text,
            image_type text,
            image_data blob)}
  end

  def add_story(story)
    begin
      @db.execute 'insert into images values(?, ?, ?, ?, ?, ?, ?, ?, ?)', story
    rescue SQLite3::Exception => e
      $logger.error('failed to write story in db: ' + e.message)
    end
  end
end




def fetch_url(url, with_type = false)
  $logger.debug('fetch_url: "' + url + '"')
  begin
    open(url) do |f|
      if f.status[0] != '200'
        $logger.error('failed: fetch_url: "' + url + '"')
        return nil
      end
      g_body = f.read
      if with_type
        return [g_body, f.metas['content-type'][0]]
      else
        return g_body
      end
    end
  rescue Exception => e
    $logger.error('fetch_url, exception: ' + e.message)
    return nil
  end
end



$feeds = []
$storage = Storage.new('news')

File.foreach("rss.txt") do |line|
  rss_url = line.strip
  rss_xml = fetch_url(rss_url)
  if rss_xml
    $feeds << rss_xml
  end
end

$feeds.each do |feed_xml|
  feed = Feedzirra::Feed.parse(feed_xml)
  feed.entries[0...MAX_ENTRIES].each do |entry|
    html_text = fetch_url(entry.url)
    html_doc = Nokogiri::HTML(html_text)
    html_doc.css('img').each do |img|
      if img['src'] and img['src'].start_with?('http://')
        $logger.debug("ask image for '#{img['src']}'")
        img = fetch_url(img['src'], true)
        if img
          img_data, img_type = *img
          img_img = Dimensions(StringIO.new(img_data))
          if img_img and ((img_img.height and img_img.height < 256) or (img_img.width and img_img.width < 256))
            $logger.debug("rejecting small image #{img_img.dimensions}")
          else
            img_data_url = 'data:' + img_type + ';base64,' + Base64.strict_encode64(img_data)
            story = [nil, feed.title, feed.url, nil, entry.url, entry.title, entry.summary, img_type, img_data_url]
            $storage.add_story(story)
          end
        end
      end
    end
  end
end
